[jayjay19630] iP

jayjay19630 opened the pull request:
# GLaDOS Command Line Interface

> You are kidding me. Did you just stuff that Aperture Science thing-we-don't-know-what-it-does into an Aperture Science Emergency Intelligence Incinerator? - GLaDOS

This is a greenfield Java project named after the fictional character named after the video game series _Portal_. Given below are instructions on how to use it.

- use the `add` command
- use some other commands
- have fun with GLaDOS!

## Setting up in Intellij

Prerequisites: JDK 17, update Intellij to the most recent version.

1. Open Intellij (if you are not in the welcome screen, click `File` > `Close Project` to close the existing project first)
2. Open the project into Intellij as follows:
   1. Click `Open`.
   2. Select the project directory, and click `OK`.
   3. If there are any further prompts, accept the defaults.
3. Configure the project to use **JDK 17** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br> In the same dialog, set the **Project language level** field to the `SDK default` option.
4. After that, locate the `src/main/java/Glados.java` file, right-click it, and choose `Run Glados.main()`
## Future Plans
- [ ] Create more filters to find task items.
- [ ] ~~Add secret commands~~.
- [ ] Refactor everything into React :anguished:.  This would look something like:
```typescript
import * from React

const Glados = () => {
    return <div>Hello User!</div>
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jayjay19630 committed:
feat: create greet and exit interface with glados

jayjay19630 committed:
feat: create echo function

jayjay19630 committed:
feat: format cli lines

jayjay19630 committed:
feat: create add and list functions

jayjay19630 committed:
feat: create task class and mark feature

jayjay19630 committed:
feat: create subclasses for task

jayjay19630 committed:
feat: add date string to event and deadline

jayjay19630 committed:
fix: format lines according to standard

jayjay19630 committed:
chore: create ui test using i/o redirection

jayjay19630 committed:
chore: change task array to array list

jayjay19630 committed:
feat: create exceptions for adding and unknown commands

jayjay19630 committed:
feat: create delete handler and exception

jayjay19630 committed:
fix: edit readme

jayjay19630 committed:
refactor: follow coding standard for switch

jayjay19630 committed:
chore: refactor into taskmanager

jayjay19630 committed:
feat: create save feature

jayjay19630 committed:
feat: create abstraction for storage class

jayjay19630 committed:
feat: create abstraction for parsing tasks

jayjay19630 committed:
fix: return array list instead of null in storage util class

jayjay19630 committed:
refactor: make tasknotfound exception more general

jayjay19630 committed:
feat: abstract ui into its own class

jayjay19630 committed:
refactor: utilise ui method for errors

jayjay19630 committed:
refactor: separate run and main

jayjay19630 committed:
refactor: divide classes into packages

jayjay19630 committed:
Merge branch 'add-gradle-support' of github.com:jayjay19630/ip into add-gradle-support

jayjay19630 committed:
feat: link gradle build to glados class

jayjay19630 committed:
feat: create tests for parser

jayjay19630 committed:
feat: create tests for task list

jayjay19630 committed:
feat: save dates as java LocalDates

jayjay19630 committed:
refactor: fix codestyle for imports

jayjay19630 committed:
feat: add javadoc for exception constructors

jayjay19630 committed:
feat: add javadocs to public methods

jayjay19630 committed:
chore: fix coding style

jayjay19630 committed:
refactor: move everything under cs2103w10.glados package

jayjay19630 committed:
feat: add javadocs to task.java

jayjay19630 committed:
Merge branch 'branch-A-CodingStandard'

jayjay19630 committed:
fix: create parser for commands

jayjay19630 committed:
refactor: add parsing for first bye input

jayjay19630 committed:
feat: add find command

jayjay19630 committed:
Merge branch 'branch-Level-9'

jayjay19630 committed:
feat: create checkstyle configs

jayjay19630 committed:
fix: do checkstytle and fix

jayjay19630 committed:
Merge branch 'A-CheckStyle'

foothaddeus commented on src/main/java/cstwooneohthree/glados/utils/Ui.java:
```diff
@@ -0,0 +1,154 @@
+     *
+     * @param e Exception to be printed.
+     */
+    public static void error(Exception e) {
+        System.out.println(
+                HORIZONTAL_LINE
+                + "\nGLaDOS: " + e.getMessage() + "\n"
+                + HORIZONTAL_LINE);
+    }
```
Perhaps this could be named "printError" instead to make it a verb?

jayjay19630 replied:
Yes, that makes sense. I will make sure to change to printError instead.

foothaddeus commented on src/main/java/cstwooneohthree/glados/utils/Parser.java:
```diff
@@ -0,0 +1,113 @@
+     * @throws ArgumentsNotFoundException If arguments do not exist.
+     */
+    public static String parseArguments(String input) throws ArgumentsNotFoundException {
+        String[] parsedArgs = input.split(" ", 2);
```
It's nice that your variable names are also sufficiently descriptive!

foothaddeus commented on src/main/java/cstwooneohthree/glados/tasks/Task.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return Task description in save format.
+     */
+    public abstract String getSaveFormat();
```
I really like that your method names in general are phrased as verbs and detail their functions clearly!

foothaddeus commented on src/main/java/cstwooneohthree/glados/tasks/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+public class TaskList {
+
+    /* ArrayList to keep track of all tasks */
+    private ArrayList<Task> tasks;
+    /* Integer to keep track of tasks length */
+    private int listIndex;
```
Your variable names clearly distinguish between single-valued and multi-valued variables, which is great!

tzuennn replied:
The naming of tasks length to be listIndex can be a bit confusing. Index usually represents an integer value indicating a particular position in a list/array. Else agree with the above, good distinction between single and multi-valued variables!

foothaddeus reviewed (commented):
Overall, brilliantly done! Everything seems great!

jayjay19630 committed:
refactor: abstract ui based on ui type

jayjay19630 committed:
feat: create gui for glados

jayjay19630 committed:
Merge branch 'branch-Level-10'

tzuennn commented on src/main/java/cstwooneohthree/glados/Glados.java:
```diff
@@ -0,0 +1,135 @@
+
+        // Run a loop indefinitely to execute commands until program exits through bye
+        while (true) {
+            String input = sc.nextLine().toLowerCase();
```
I liked that you handled user input, ensuring the program works with case-insensitive command.

tzuennn commented on src/main/java/cstwooneohthree/glados/tasks/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+
+        return new String[]{task.toString(), String.valueOf(listIndex)};
+    }
+
```
Nice way of handling index out of range!

tzuennn commented on src/main/java/cstwooneohthree/glados/Glados.java:
```diff
@@ -0,0 +1,135 @@
+                executeCommand(input);
+            }
+        }
+        sc.close();
```
Rather than having to explicitly close the scanner, you can try using try-with block. This will attempt to automatically close the scanner once the block is left.

jayjay19630 replied:
That's the first time I've heard about try-with blocks. Thanks for letting me know, they seem quite useful! I'll try to use it here to make the code look simpler.

tzuennn commented on src/main/java/cstwooneohthree/glados/utils/Ui.java:
```diff
@@ -0,0 +1,199 @@
+            + "░▒▓█▓▒░░▒▓█▓▒░▒▓█▓▒░      ░▒▓█▓▒░░▒▓█▓▒░▒▓█▓▒░░▒▓█▓▒░▒▓█▓▒░░▒▓█▓▒░      ░▒▓█▓▒░ \n"
+            + "░▒▓█▓▒░░▒▓█▓▒░▒▓█▓▒░      ░▒▓█▓▒░░▒▓█▓▒░▒▓█▓▒░░▒▓█▓▒░▒▓█▓▒░░▒▓█▓▒░      ░▒▓█▓▒░ \n"
+            + " ░▒▓██████▓▒░░▒▓████████▓▒░▒▓█▓▒░░▒▓█▓▒░▒▓███████▓▒░ ░▒▓██████▓▒░░▒▓███████▓▒░  \n"
+            + "\n";
```
This is kinda sick 😮 .

tzuennn commented on src/main/java/cstwooneohthree/glados/utils/Ui.java:
```diff
@@ -0,0 +1,199 @@
+                        + task + "\n" 
+                        + "Now you have " + index + (Integer.parseInt(index) == 1 ? " task." : " tasks.")
+                        + "\n";
+
```
Good job considering even the most minor stuff, singular for a task and plural for multiple tasks!

tzuennn commented on src/main/java/cstwooneohthree/glados/utils/Storage.java:
```diff
@@ -0,0 +1,101 @@
+            String line;
+            while ((line = reader.readLine()) != null) {
+                String[] inputs = line.split(" \\| ");
+                String taskType = inputs[0];
```
Naming can be misleading. I would think of the enums DEADLINE, TODO, EVENT since you named it as taskType. Generally, when we talk about "task type," we think of something more descriptive and type-safe, like enums. Perhaps formattedTaskType would be a better choice.

tzuennn commented on src/main/java/cstwooneohthree/glados/utils/Storage.java:
```diff
@@ -0,0 +1,101 @@
+                case "[E]":
+                    task = new Event(inputs[2], LocalDate.parse(inputs[3]), LocalDate.parse(inputs[4]));
+                    break;
+                default:
```
Default case is present but is not handling any logic. I understand that almost all exceptions and errors are already being handled by other classes. But it would still be good to take care of unexpected exception that we might miss out. Example:
```java
default:
    throw new IllegalArgumentException("Invalid task type: " + taskType);
```

tzuennn reviewed (commented):
Overall good job! Super clear separation of concerns.

AppleJem commented on README.md:
```diff
@@ -1,24 +1,36 @@
+
+   -----------------------------------------------------------------------------
+
+   Hello, welcome to the Aperture Science computer-aided enrichment center! My name is:
```
Goddam that's pretty cool, I too am a Portal 2 fan

AppleJem commented on src/main/java/cstwooneohthree/glados/Glados.java:
```diff
@@ -0,0 +1,135 @@
+    public String executeCommand(String input) {
+        try {
+            String query = Parser.parseCommand(input);
+            switch (query) {
```
Oh great use of switch here, I used one megalong if else oops

AppleJem commented on src/main/java/cstwooneohthree/glados/Glados.java:
```diff
@@ -0,0 +1,135 @@
+        return ui.add(res[0], res[1]);
+    }
+
+    private String delete(int index) throws TaskNotFoundException {
```
Nitpick, might be clearer to name it deleteTask

AppleJem commented on src/main/java/cstwooneohthree/glados/utils/Storage.java:
```diff
@@ -0,0 +1,101 @@
+                }
+            }
+            return taskList;
+        } catch (DateTimeParseException e) {
```
Great that you have different error statements to tell the user what exactly went wrong

AppleJem commented on src/main/java/cstwooneohthree/glados/utils/Ui.java:
```diff
@@ -0,0 +1,199 @@
+            System.out.println(HORIZONTAL_LINE + message + HORIZONTAL_LINE);
+            return null;
+        } else {
+            return "I have removed the following task from the list: " 
```
Kudos on considering the plural hahaha

AppleJem commented on src/main/java/cstwooneohthree/glados/utils/Ui.java:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @param taskList Task list to be printed.
+     */
+    public String list(ArrayList<Task> taskList, boolean shouldFindMatching) {
```
I thought shouldFindmatching might be slightly ambiguous? Not quite clear what it does. Maybe explain in the docstring or a comment

AppleJem reviewed (commented):
That's a really good job! Learned a lot from reading your code

jayjay19630 committed:
Add task type assertions for tasklist and parser.

Methods in tasklist and parser expect defined task types such as Todo,
Deadline and Event. Encountering another enum value would not be
expected.

The corresponding assertions define that the tasktypes must be from
one of those three enum values.

jayjay19630 committed:
feat: create assertion for ui.java

jayjay19630 committed:
refactor: use streams in task list

jayjay19630 committed:
Merge pull request #2 from jayjay19630/branch-A-Streams

refactor: Use streams in task list

jayjay19630 committed:
Merge pull request #1 from jayjay19630/branch-A-Assertions

feat: Add assertions

jayjay19630 committed:
refactor: change shouldmatch to isFilteredList

jayjay19630 committed:
refactor: use default branch in load tasks

jayjay19630 committed:
refactor: Create better spacing

This commit makes it so that some ideas are better spaced apart
for readability.

Functionally, nothing has changed about the code.

jayjay19630 committed:
Merge pull request #3 from jayjay19630/branch-A-CodeQuality

refactor: Improve code quality

jayjay19630 committed:
feat: create yml file for workflow

jayjay19630 committed:
refactor: fix checkstyle issues

jayjay19630 committed:
refactor: remove io redirection test in yml

jayjay19630 committed:
Merge pull request #4 from jayjay19630/branch-A-CI

feat: create CI pipeline

jayjay19630 committed:
feat: Add ability to update task description

This commit gives users the ability to update task description
by specifying the index of the task and the new description.

This feature also comes with new exceptions that might be thrown
if the index is out of range of the tasklist.

jayjay19630 committed:
Merge pull request #5 from jayjay19630/branch-C-Update

feat: Add ability to update task description

jayjay19630 committed:
fix: resolve chcekstyle issues

jayjay19630 committed:
feat: add glados background and reposition text box

jayjay19630 committed:
feat: create transparent glados dialog box

jayjay19630 committed:
feat: add font to glados dialog

jayjay19630 committed:
feat: create transition between scenes

jayjay19630 committed:
fix: set max width and min height

jayjay19630 committed:
feat: add more personality to glados

jayjay19630 committed:
Merge branch 'branch-A-Personality'

jayjay19630 committed:
feat: add ui.png to docs

jayjay19630 committed:
feat: create product description and quick start in docs

jayjay19630 committed:
feat: create docs for add and delete

jayjay19630 committed:
fix: change header style

jayjay19630 committed:
feat: create rest of docs

jayjay19630 committed:
fix: complete table of contents

jayjay19630 committed:
fix: resolve checkstyle issues

jayjay19630 committed:
fix: add more info to docs

jayjay19630 committed:
fix: create clarification in product description

jayjay19630 committed:
fix: add dependencies for javafx

jayjay19630 committed:
feat: add docs for echo

jayjay19630 committed:
feat: create stubs for tasklist unit tests

jayjay19630 committed:
feat: create more tests for deleting tasks

jayjay19630 committed:
Merge branch 'branch-A-MoreTesting'
